homeassistant/components/bmw_connected_drive/__init__.py

```diff
@@ -45,48 +52,155 @@
+                    DOMAIN, context={"source": "import"}, data=entry_config
+                )
+            )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up BMW Connected Drive from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # Convert data to dict to remove settings that are now stored as options
+    entry.data = dict(entry.data)
+    default_options = {
+        CONF_READ_ONLY: entry.data.pop(CONF_READ_ONLY, False),
+        CONF_USE_LOCATION: False,
+    }
 
-    hass.data[DOMAIN] = accounts
+    # Create options based on user input if no options are stored
+    if list(entry.options) != list(default_options):
+        default_options.update(entry.options)
+        entry.options = default_options
+        hass.config_entries.async_update_entry(entry, options=entry.options)
+
+    try:
+        account = await hass.async_add_executor_job(
+            setup_account, entry, hass, entry.data[CONF_USERNAME]
+        )
+        await hass.async_add_executor_job(account.update)
+    except Exception as ex:
+        raise ConfigEntryNotReady from ex
 
-    def _update_all(call) -> None:
+    hass.data[DOMAIN][entry.entry_id] = account
+
+    async def _async_update_all(service_call=None):
         """Update all BMW accounts."""
-        for cd_account in hass.data[DOMAIN]:
+        await hass.async_add_executor_job(_update_all)
+
+        return True
+
+    def _update_all() -> None:
+        """Update all BMW accounts."""
+        for cd_account in list(hass.data[DOMAIN].values()):
             cd_account.update()
 
+    # Add update listener for config entry changes (options)
+    if weakref.ref(update_listener) not in entry.update_listeners:
+        entry.add_update_listener(update_listener)
+
     # Service to manually trigger updates for all accounts.
-    hass.services.register(DOMAIN, SERVICE_UPDATE_STATE, _update_all)
+    hass.services.async_register(DOMAIN, SERVICE_UPDATE_STATE, _async_update_all)
 
-    _update_all(None)
+    await _async_update_all()
 
-    for component in BMW_COMPONENTS:
-        discovery.load_platform(hass, component, DOMAIN, {}, config)
+    for platform in BMW_PLATFORMS:
+        if platform != NOTIFY_DOMAIN:
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, platform)
+            )
+
+    hass.async_create_task(
+        discovery.async_load_platform(hass, NOTIFY_DOMAIN, DOMAIN, {}, entry.data)
+    )
 
     return True
 
 
-def setup_account(account_config: dict, hass, name: str) -> "BMWConnectedDriveAccount":
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in BMW_PLATFORMS
+                if component != NOTIFY_DOMAIN
```

Comment:
Because we load NOTIFY_DOMAIN in 121-123 with async_load_platform instead of async_forward_entry_setup. Therefore all services created by NOTIFY_DOMAIN need to be handled accordingly afterwards?

It's been a while since I've written that code and honestly just copy&pasted what works. 
What we want is a NOTIFY service for each car, and if an account has multiple cars, then each car should have its own.

If there is a better way to do it, please advise.

Reply:
The `notify` platform doesn't seem to support `config_entries` yet (https://github.com/home-assistant/core/blob/dev/homeassistant/components/ios/__init__.py#L262). 
I would keep it in `BMW_PLATFORMS` to make clear we're using it. But if you prefer, I can remove it from `BMW_PLATFORMS` and save two `if component != NOTIFY_DOMAIN`.